src/software/backend/replay_logging/replay_reader.cpp

```diff
@@ -0,0 +1,137 @@
+    bool result = google::protobuf::util::ParseDelimitedFromCodedStream(
+        &msg, coded_input.get(), nullptr);
+    if (!result)
+    {
+        throw std::invalid_argument("Failed to parse protobuf from file " +
+                                    file_path.string());
+    }
+    close(fd);
+    return msg;
+}
+
+
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const std::vector<T>& set)
+{
+    os << '{';
+    for (const auto& val : set)
+    {
+        os << val << ' ';
+    }
+    os << '}';
+    return os;
+}
+
+ReplayReader::ReplayReader(const std::string& _replay_dir)
+    : cur_msg_idx(0), replay_dir(_replay_dir)
+{
+    if (!fs::exists(replay_dir) || !fs::is_directory(replay_dir))
+    {
+        throw std::invalid_argument(replay_dir.string() +
+                                    "does not exist or is not a directory!");
+    }
+
+    if (fs::is_empty(replay_dir))
+    {
+        throw std::invalid_argument(replay_dir.string() + " is empty!");
+    }
+
+    // we assume chunk indices are monotonically increasing, but not necessarily starting
+    // at 0; this is so that we can trivially 'crop' replays
+    fs::directory_iterator dir_it(replay_dir);
+    // filter the filenames in the replay_logging out directory to only numeric filenames
+    std::vector<size_t> chunk_indices;
+    std::for_each(fs::begin(dir_it), fs::end(dir_it),
+                  [&chunk_indices](const fs::directory_entry& dir_entry) {
+                      try
+                      {
+                          auto fname_num = std::stoul(dir_entry.path().filename());
+                          chunk_indices.emplace_back(fname_num);
+                      }
+                      catch (std::invalid_argument&)
+                      {
+                      };
+                  });
+
+    if (chunk_indices.size() == 0)
+    {
+        throw std::invalid_argument(
+            replay_dir.string() +
+            " contains no files with numerical names! It is not a valid replay_logging directory!");
+    }
+
+    std::sort(chunk_indices.begin(), chunk_indices.end());
+    cur_chunk_idx   = *chunk_indices.begin();
+    max_chunk_idx   = *chunk_indices.rbegin();
+    auto chunk_path = replay_dir / std::to_string(cur_chunk_idx);
+
+    cur_chunk.CopyFrom(readDelimitedReplayProtobufFile(chunk_path));
+}
+
+std::optional<SensorMsg> ReplayReader::getNextMsg()
+{
+    if (!hasNextMsg())
+    {
+        return std::nullopt;
+    }
+
+    if (cur_msg_idx >= cur_chunk.replay_msgs_size())
+    {
+        nextChunk();
+    }
+
+    auto ret = cur_chunk.replay_msgs(cur_msg_idx);
+    cur_msg_idx++;
+    return ret;
+}
+
+bool ReplayReader::hasNextMsg() const
+{
+    return !(cur_chunk_idx == max_chunk_idx &&
+             cur_msg_idx == cur_chunk.replay_msgs_size());
+}
+
+void ReplayReader::nextChunk()
+{
+    cur_chunk_idx++;
+    auto chunk_path = replay_dir / std::to_string(cur_chunk_idx);
+
+    if (!fs::is_regular_file(chunk_path))
```

Comment:
You mentioned "cropping" replays above, but the only termination condition here seems to be there are no more valid files left. If we "cropped" a replay to just play the middle chunks, it seems we would still play the end chunks because the files exist?

Reply:
The idea was that we could have cropped replays just by selecting some 'chunks' and then copypasting them into another directory.

Reply:
Oh I didn't realize that would involve copying. This should be fine for now then.